### PR TITLE
Provide an offline engine API

### PR DESCRIPTION
**Pull request by ByronHsu**
## Motivation

This PR is to support "Add APIs for using the inference engine in a single script without launching a separate server" in https://github.com/sgl-project/sglang/issues/1487. This is a simplified version of https://github.com/sgl-project/sglang/pull/1127 @JianyuZhan  where I reuse most of the existing code.

### Context

The current SRT server consists of an HTTP server and the SRT engine.

1. HTTP server: A FastAPI server that routes requests to the engine.
2. SRT engine:
    1. Tokenizer Manager: Tokenizes the requests and sends them to the controller.
    2. Controller (subprocess): Receives requests from the Tokenizer Manager, schedules batches, forwards them, and sends the output tokens to the Detokenizer Manager.
    3. Detokenizer Manager (subprocess): Detokenizes the output tokens and sends the result back to the Tokenizer.

HTTP server and Tokenizer Manager are both running in the main process, but there is no way to decouple them and only instantiate Tokenizer Manager.

### Decouple SRT engine and HTTP server

This PR introduces **SRT engine**  by decoupling `launch_server` to `launch_server` and `launch_engine`. 

`launch_server`: `launch_engine` + HTTP server creation, used by SRT Runtime and standalone server.
`launch_engine`: SRT Engine creation, used by SRT engine.

### New public API: `Engine`

Uplift Engine to the top level, so users can easily call with `sgl.Engine`

### Engine Usage Example

Same settings as [vllm](https://docs.vllm.ai/en/latest/getting_started/examples/offline_inference.html) but use SRT Engine.

```python
import sglang as sgl

# Sample prompts.
prompts = [
    "Hello, my name is",
    "The president of the United States is",
    "The capital of France is",
    "The future of AI is",
]
# Create a sampling params object.
sampling_params = {"temperature": 0.8, "top_p": 0.95}

# Create an LLM.
llm = sgl.Engine(model_path="meta-llama/Meta-Llama-3.1-8B-Instruct")

outputs = llm.generate(prompts, sampling_params)
# Print the outputs.
for prompt, output in zip(prompts, outputs):
    print("===============================")
    print(f"Prompt: {prompt}\nGenerated text: {output['text']}")

```

```
===============================
Prompt: Hello, my name is
Generated text:  Alistair. I am an independent game developer.

I am currently working on a virtual reality game that focuses on exploring the world and discovering new things. The game has a unique art style that combines elements of both photorealism and cartoonish art. The game will have no story, but rather focus on the exploration of the world and the various things that can be found within.

I am looking for a 3D artist who can help me bring my vision to life. The ideal candidate will have experience in creating environments, characters, and objects in Unity. They should also have a strong
===============================
Prompt: The president of the United States is
Generated text:  Donald Trump. The president of India is Narendra Modi. The president of China is Xi Jinping. The president of Russia is Vladimir Putin. The president of France is Emmanuel Macron. The president of the United Kingdom is Boris Johnson. The president of Canada is Justin Trudeau. The president of Australia is Scott Morrison. The president of Japan is Shinzo Abe. The president of South Korea is Moon Jae-in. The president of Brazil is Jair Bolsonaro. The president of Colombia is Ivan Duque. The president of Argentina is Mauricio Macri
===============================
Prompt: The capital of France is
Generated text:  Paris. Paris is one of the world’s leading cities and is known for its history, art, culture, architecture, and gastronomy. Paris is located in the north-central region of France and is surrounded by the Seine River, which flows through the city. The city is divided into twenty districts or arrondissements, each with its own unique character and charm. Some of the most famous landmarks in Paris include the Eiffel Tower, the Louvre Museum, Notre-Dame Cathedral, the Arc de Triomphe, and the Champs-Élysées. Paris is also known
===============================
Prompt: The future of AI is
Generated text:  bright, and its potential to revolutionize various industries is immense. Here are some ways in which AI can impact the future of healthcare:

1. Personalized Medicine: AI can help in developing personalized treatment plans for patients based on their genetic makeup, medical history, and lifestyle factors. This can lead to more effective treatments and better outcomes for patients.
2. Diagnosis and Treatment: AI-powered tools can help healthcare providers diagnose diseases more accurately and quickly. They can also help in selecting the most appropriate treatment options based on the patient's condition.
3. Drug Development: AI can help in acceler
```

### Discussion

One caveat is that we construct `ServerArgs`, but the HTTP server related args will not be used. I think this is ok because  ServerArgs is the superset of Engine Args, so it can cover everything.

## Testing

Add `test_srt_engine.py`, which runs batch inference and assert the answer.

## TODO

1. Add async generate
2. Add encode

## Checklist

- [x] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [x] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [x] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

**Comment by zhyncs** [comment]
I think `async generate` is needed. QQ What does `add decode` in todo 2 refer to?

**Comment by ByronHsu** [comment]
@zhyncs oh i mean `encode` like the one in `Runtime`. It was a typo

**Comment by ByronHsu** [comment]
@zhyncs i can do the async gen in the next PR

File: test/lang/test_srt_engine.py
Comment (Ying1123) [inline test/lang/test_srt_engine.py:1]: The unit test needs to be added into the `minimal` suite in `test/srt/run_suite.py`.

File: test/lang/test_srt_engine.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    def test_generate(self):
+        prompts = [
+            "The capital of China is",
+            "The square root of 144 is",
+        ]
+
+        outputs = self.engine.generate(prompts, {"max_new_tokens": 128})
+
+        assert "Beijing" in outputs[0]["text"]
+        assert "12" in outputs[1]["text"]
```
Comment (Ying1123) [inline test/lang/test_srt_engine.py]: Can we add a more strict comparison? For example, asserting a long output is exactly the same as the one obtained by using the old Runtime interface.

File: python/sglang/api.py [hunk 2]
```diff
@@ -40,6 +40,14 @@ def Runtime(*args, **kwargs):
     return Runtime(*args, **kwargs)
 
 
+def Engine(*args, **kwargs):
+    # Avoid importing unnecessary dependency
+    os.environ["TF_CPP_MIN_LOG_LEVEL"] = "3"
+    from sglang.srt.server import Engine
```
Comment (merrymercy) [inline python/sglang/api.py]:
```suggestion
    os.environ["TF_CPP_MIN_LOG_LEVEL"] = "3"
    # Avoid importing unnecessary dependency
    from sglang.srt.server import Engine
```

File: python/sglang/srt/server.py [hunk 3]
```diff
@@ -355,6 +359,29 @@ def launch_server(
...
+
+    1. HTTP server: A FastAPI server that routes requests to the engine.
+    2. SRT engine:
+        1. Tokenizer Manager: Tokenizes the requests and sends them to the controller.
+        2. Controller (subprocess): Receives requests from the Tokenizer Manager, schedules batches, forwards them, and sends the output tokens to the Detokenizer Manager.
+        3. Detokenizer Manager (subprocess): Detokenizes the output tokens and sends the result back to the Tokenizer Manager.
+
+    Note:
+    1. The HTTP server and Tokenizer Manager both run in the main process.
+    2. Inter-process communication is done through TCP (each process uses a different port) via the ZMQ library.
```
Comment (merrymercy) [inline python/sglang/srt/server.py]: TCP -> IPC.

We removed TCP recently in this PR https://github.com/sgl-project/sglang/pull/1566.

File: examples/frontend_language/usage/srt_engine.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import sglang as sgl
```
Comment (merrymercy) [inline examples/frontend_language/usage/srt_engine.py]: this is unrelated to the frontend language. Can we move this under `examples/runtime`?

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: python/sglang/srt/server.py (hunk 3 above)
Comment (ByronHsu) [reply python/sglang/srt/server.py]: Realized that i was developing on v0.3.2. The engine is now failing on the master HEAD as @Ying1123 found out. Let me dig into the issue

File: examples/frontend_language/usage/srt_engine.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+# Sample prompts.
```
Comment (Ying1123) [inline examples/frontend_language/usage/srt_engine.py]: The code needs to be wrapped in `if __name__ == "__main__"` to make it work. Because the engine use `spawn` to create subprocesses.
Please add this hint into the example.

File: python/sglang/srt/server.py [hunk 6]
```diff
@@ -626,3 +653,49 @@ def encode(
...
+            logprob_start_len=logprob_start_len,
+            top_logprobs_num=top_logprobs_num,
+            lora_path=lora_path,
+        )
+
+        async def async_generate():
+            return await generate_request(obj, None)
+
+        # make it synchronous
+        return asyncio.run(async_generate())
```
Comment (Ying1123) [inline python/sglang/srt/server.py]: Why don't just
```suggestion
        # make it synchronous
        return asyncio.run(generate_request(obj, None))
```

Comment (ByronHsu) [reply python/sglang/srt/server.py]: brilliant

File: examples/frontend_language/usage/srt_engine.py (hunk 5 above)
Comment (ByronHsu) [reply examples/frontend_language/usage/srt_engine.py]: thanks for catching this!!

File: examples/frontend_language/usage/srt_engine.py (hunk 4 above)
Comment (ByronHsu) [reply examples/frontend_language/usage/srt_engine.py]: moved

File: test/lang/test_srt_engine.py (hunk 1 above)
Comment (ByronHsu) [reply test/lang/test_srt_engine.py]: The consistency check is a bit tricky because:

1. It's not possible to pass a random seed from the client side; it can only be set during engine/runtime initialization.
2. When using the runtime, the client runs a warmup thread, which adds an extra dummy request for "The capital city of France". This causes the random seed to advance.

As a result, both scenarios have different random seeds when processing "what is AGI?".

Engine:
```
GenerateReqInput(text='what is AGI?',
```

Runtime:
```
GenerateReqInput(text='The capital city of France is'
GenerateReqInput(text='what is AGI?',
```

This difference in random seed states leads to potential inconsistencies when handling the same input ("what is AGI?") between the Engine and Runtime implementations.

What if we make the test case harder? or anyway to fix random seed per request?  cc @merrymercy if you have any inputs

File: python/sglang/srt/server.py [hunk 7]
```diff
@@ -355,6 +359,29 @@ def launch_server(
...
 
+
+def launch_server(
+    server_args: ServerArgs,
+    pipe_finish_writer: Optional[mp.connection.Connection] = None,
+):
+    """
+    Launch SRT (SGLang Runtime) Server
+
+    The SRT server consists of an HTTP server and the SRT engine.
```
Comment (merrymercy) [inline python/sglang/srt/server.py:370]: nice comment!

File: python/sglang/srt/server.py [hunk 8]
```diff
@@ -355,6 +359,29 @@ def launch_server(
...
+
+    1. HTTP server: A FastAPI server that routes requests to the engine.
+    2. SRT engine:
+        1. Tokenizer Manager: Tokenizes the requests and sends them to the controller.
+        2. Controller (subprocess): Receives requests from the Tokenizer Manager, schedules batches, forwards them, and sends the output tokens to the Detokenizer Manager.
```
Comment (merrymercy) [inline python/sglang/srt/server.py]: We did some renaming. Controller -> Scheduler

File: test/lang/run_suite.py [hunk 9]
```diff
@@ -4,7 +4,7 @@
 from sglang.test.test_utils import run_unittest_files
 
 suites = {
-    "minimal": ["test_srt_backend.py", "test_openai_backend.py"],
+    "minimal": ["test_srt_backend.py", "test_openai_backend.py", "test_srt_engine.py"],
```
Comment (merrymercy) [inline test/lang/run_suite.py]: This engine is not related to the frontend, so please move this test under `test/srt/`

File: test/lang/test_srt_engine.py (hunk 1 above)
Comment (ByronHsu) [reply test/lang/test_srt_engine.py]: discussed offline, setting `sampling_backend='pytorch'` and temp = 0 can be deterministic regardless of the random seed

File: test/lang/run_suite.py (hunk 9 above)
Comment (ByronHsu) [reply test/lang/run_suite.py]: done

File: examples/runtime/srt_engine.py [hunk 10]
```diff
@@ -0,0 +1,23 @@
+import sglang as sgl
+
+# The __main__ condition is necessary here because we use "spawn" to create subprocesses
+# Spawn starts a fresh program every time, if there is no __main__, it will run into infinite loop to keep spawning processes from sgl.Engine
+if __name__ == "__main__":
+    # Sample prompts.
+    prompts = [
```
Comment (merrymercy) [inline examples/runtime/srt_engine.py:6]: wrap this into a function `main`?

```
def main():
    prompts = ...

if __name__ == "__main__":
    main()
```

**Review by merrymercy — APPROVED** [review]

File: examples/runtime/srt_engine.py (hunk 10 above)
Comment (ByronHsu) [reply examples/runtime/srt_engine.py:6]: changed

**Comment by ByronHsu** [comment]
Please don't merge now. Consistency test is failing on H100 in CI, but passing on my A100

File: test/srt/test_srt_engine.py [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+
+class TestSRTBackend(unittest.TestCase):
+
+    def test_engine_runtime_consistency(self):
+        prompt = "Today is a sunny day and I like"
+        model_path = DEFAULT_MODEL_NAME_FOR_TEST
+
+        engine = sgl.Engine(model_path=model_path, random_seed=42)
+        out1 = engine.generate(prompt, {"temperature": 0})["text"]
```
Comment (merrymercy) [inline test/srt/test_srt_engine.py]:
```suggestion
        out1 = engine.generate(prompt, {"temperature": 0, "max_new_tokens": 8})["text"]
```

File: test/srt/test_srt_engine.py [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+        engine.shutdown()
+
+        runtime = sgl.Runtime(model_path=model_path, random_seed=42)
+        out2 = json.loads(runtime.generate(prompt, {"temperature": 0}))["text"]
```
Comment (merrymercy) [inline test/srt/test_srt_engine.py]:
```suggestion
        out2 = json.loads(runtime.generate(prompt, {"temperature": 0, "max_new_tokens": 8}))["text"]
```

Comment (ByronHsu) [reply test/srt/test_srt_engine.py]: the results are constantly identical on my A100 hmm

File: test/srt/test_srt_engine.py [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+        prompt = "Today is a sunny day and I like"
+        model_path = DEFAULT_MODEL_NAME_FOR_TEST
+
+        engine = sgl.Engine(model_path=model_path, random_seed=42)
+        out1 = engine.generate(prompt, {"temperature": 0, "max_new_tokens": 8})["text"]
+        engine.shutdown()
+
+        runtime = sgl.Runtime(model_path=model_path, random_seed=42)
+        out2 = json.loads(runtime.generate(prompt, {"temperature": 0, "max_new_tokens": 8}))["text"]
+        runtime.shutdown()
```
Comment (merrymercy) [inline test/srt/test_srt_engine.py]:
```suggestion
        sampling_params = {"temperature": 0, "max_new_tokens": 8}

        engine = sgl.Engine(model_path=model_path, random_seed=42)
        out1 = engine.generate(prompt, sampling_params)["text"]
        engine.shutdown()

        runtime = sgl.Runtime(model_path=model_path, random_seed=42)
        out2 = json.loads(runtime.generate(prompt, sampling_params))["text"]
        runtime.shutdown()
```

File: python/sglang/srt/server.py [hunk 14]
```diff
@@ -626,3 +652,46 @@ def encode(
...
+    SRT Engine without an HTTP server layer.
+
+    This class provides a direct inference engine without the need for an HTTP server. It is designed for use cases where
+    launching the HTTP server adds unnecessary complexity or overhead,
+    """
+
+    def __init__(self, *args, **kwargs):
+
+        # before python program terminates, call shutdown implicitly. Therefore, users don't have to explicitly call .shutdown()
+        atexit.register(self.shutdown)
```
Comment (merrymercy) [inline python/sglang/srt/server.py:668]: 1. Can we do this for `Runtime` as well?
2. Is it better than `__delete__`?

**Review by merrymercy — APPROVED** [review]

**Comment by imadoualid** [comment]
hey guys i'm getting the AttributeError: module 'sglang' has no attribute 'Engine' on sgl '0.3.2' still not in prod ?

**Comment by ByronHsu** [comment]
@imadoualid the changes should be in the main HEAD if you can build from source.
